### DISCUSSION TITLE
UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`

### DISCUSSION THREAD

**Discussion by meshidenn** [horovod/horovod#2776] [Q&A]
**Environment:**
1. Pytorch
2. Framework version:1.7.1
3. Horovod version: 0.21.1
4. MPI version: 4.1.0(or 3.1.4)
5. CUDA version: 10.1.105
6. NCCL version: _2.8.3-1
7. Python version: 3.8.3
8. OS and version: SUSE Linux Enterprise Server 12 SP4"
9. GCC version:8.3.0

**Checklist:**
1. Did you search issues to find if somebody asked this question before? Yes, I cannot find any issues about this warning.
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)? No, I think this warning is not about hang.
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)? No, I think this warning is not about docker
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)? Yes, I cannot find any issues about this warning.

**Your question:**
I saw the following warning when using horovod even if I ran my code on a single node(and also multi nodes of course). Should I care the warning and how can I deal with it?  The warning didn't appear when I use pytorch-DDP.

```
/path/to/torch/optim/lr_scheduler.py:131: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
```

I know this error is related to the order of calling optimizer.step() and lr_scheduler.step()(ref: https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate). My code snippet is following. I certainly call lr_schreduler after optimizer.step() but the warning appeared.

```
import torch
from torch.utils.data import Dataset, DataLoader
from torch.utils.data.distributed import DistributedSampler
from torch import optim
from transformers import get_linear_schedule_with_warmup
...

    hvd.init()
    torch.cuda.set_device(hvd.local_rank())
    train_model.cuda()
    optimizer = optim.AdamW(train_model.parameters(), lr=1e-5)
    scheduler = get_linear_schedule_with_warmup(optimizer, num_warmup_steps, num_train_steps)
    optimizer = hvd.DistributedOptimizer(
        optimizer, named_parameters=train_model.named_parameters(), compression=compression
    )
    optimizer._requires_update = set()

    hvd.broadcast_parameters(train_model.state_dict(), root_rank=0)
    hvd.broadcast_optimizer_state(optimizer, root_rank=0)
    start_epoch = hvd.broadcast(start_epoch, root_rank=0, name="start_epoch").item()
    start_iter = hvd.broadcast(start_iter, root_rank=0, name="start_iter").item()

    sampler = DistributedSampler(train_dataset, num_replicas=hvd.size(), rank=hvd.rank())

    for i in range(epoch):
        if i < start_epoch:
            continue
        elif i == start_epoch:
            pass
        elif i > start_epoch:
            start_iter = 0

        sampler.set_epoch(epoch)
        examples = DataLoader(
            train_dataset, batch_size=batch_size_per_gpu, pin_memory=True, sampler=sampler, drop_last=True
        )

        for j, data in tqdm(enumerate(examples), total=len(examples),):
            if j < start_iter:
                continue

            loss = train_model.forward(data)
            optimizer.zero_grad()

            if fp16:
                with amp.scale_loss(loss, optimizer) as scaled_loss:
                    scaled_loss.backward()
                    optimizer.synchronize()
                with optimizer.skip_synchronize():
                    optimizer.step()

            else:
                loss.backward()
                optimizer.step()

            scheduler.step()
...
```

Thank you.

**Comment by ghost** [comment]
Yes. I'm too facing this issue. I'm broadcasting the `lr_scheduler` to all the ranks and calling the `lr_scheduler.step()` after `optimizer.step()`.

**Comment by meshidenn** [comment]
Thank you!

Do you mean you add ` broadcast_object(lr_scheduler, root_rank=9)` as suggested [here](https://github.com/horovod/horovod/issues/2416)?
